notebooks/Threshold Optimization Postprocessing for Binary Classification.ipynb

```diff
@@ -0,0 +1,452 @@
+    "        x_tick_labels_basic.append(group)\n",
+    "    \n",
+    "        if y is not None:\n",
+    "            positive_indices[group] = sensitive_features.index[(sensitive_features == group) & (y == 1)]\n",
+    "            negative_indices[group] = sensitive_features.index[(sensitive_features == group) & (y == 0)]\n",
+    "            prob_1 = sum(y_pred[positive_indices[group]])/len(positive_indices[group])\n",
+    "            prob_0 = sum(y_pred[negative_indices[group]])/len(negative_indices[group])\n",
+    "            print(\"P[recidivism predicted | {}, recidivism]       = {}\".format(group, prob_1))\n",
+    "            print(\"P[recidivism predicted | {}, no recidivism]   = {}\".format(group, prob_0))\n",
+    "\n",
+    "            plt.bar(n_groups + 1 + 2 * index, prob_1, color=color[index])\n",
+    "            plt.bar(n_groups + 2 + 2 * index, prob_0, color=color[index])\n",
+    "            x_tick_labels_by_label.extend([\"{} recidivism\".format(group), \"{} no recidivism\".format(group)])\n",
+    "    \n",
+    "    x_tick_labels = x_tick_labels_basic + x_tick_labels_by_label\n",
+    "    plt.xticks(range(1, len(x_tick_labels)+1), x_tick_labels, rotation=45, horizontalalignment=\"right\")\n",
+    "        \n",
+    "show_proportions(X_train, sensitive_features_train, y_train, description=\"original training data:\", plot_row_index=1)\n",
+    "show_proportions(X_train, sensitive_features_train, unconstrained_predictor.predict(X_train), y_train, description=\"fairness-unaware prediction on training data:\", plot_row_index=2)\n",
+    "show_proportions(X_test, sensitive_features_test, y_test, description=\"original test data:\", plot_row_index=3)\n",
+    "show_proportions(X_test, sensitive_features_test, unconstrained_predictor.predict(X_test), y_test, description=\"fairness-unaware prediction on test data:\", plot_row_index=4)\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We notice a stark contrast in the predictions with males being a lot more likely to be predicted to get approved, similar to the original training data. However, there's even a disparity between the subgroup of males and females that are approved. When considering only the samples labeled with \"loan not repaid\" males are much more likely to be predicted to get approved than females. The test data shows a similar disparity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fairlearn.widget import FairlearnDashboard\n",
+    "\n",
+    "FairlearnDashboard(sensitive_features=sensitive_features_test.values,\n",
+    "                   true_y=y_test.values,\n",
+    "                   predicted_ys=unconstrained_predictor.predict(X_test),\n",
+    "                   class_names=None,\n",
+    "                   feature_names=X_test.columns.values.tolist(),\n",
```

Comment:
here, feature names is not the same length as sensitive features, so it misbehaves (doesn't raise an exception, but it should). Please either pass in just the array for the senitive feature column names(eg, ["race"]) or leave None

Reply:
Oh, so this should be "sensitive_feature_names"